src/Network/Freddy.hs

```diff
@@ -208,16 +206,11 @@ matchingCorrelationId correlationId msg =
 topicExchange :: Text
 topicExchange = "freddy-topic"
 
-waitForResponse :: ResponseChannelListener -> CorrelationId -> (AMQP.Message -> Bool) -> IO AMQPResponse
-waitForResponse eventChannelListener correlationId predicate = do
-  amqpResponse <- BC.readBChan eventChannelListener
+waitForResponse :: ResponseChannelListener -> (AMQP.Message -> Bool) -> IO (Maybe AMQP.PublishError, AMQP.Message)
+waitForResponse eventChannelListener predicate = do
+  (error, msg) <- BC.readBChan eventChannelListener
 
-  case amqpResponse of
-    Right msg ->
-      if predicate msg then
-        return amqpResponse
-      else
-        waitForResponse eventChannelListener correlationId predicate
-    Left error ->
-      -- TODO: Check routing key. This is needed when having multiple threads.
-      return amqpResponse
+  if predicate msg then
+    return (error, msg)
+  else
+    waitForResponse eventChannelListener predicate
```

Comment:
Didn't find anything https://hackage.haskell.org/package/broadcast-chan-0.1.0/docs/Control-Concurrent-BroadcastChan.html . Are there performance concerns? It looked like haskell can nicely optimize it.

Currently DWP 10K messages takes 3.8 seconds in haskell and 4.8s in ruby. Although ruby also includes json dumping & parsing.